README.md
**NOTE**: This is a beta! Some cursors need to be touched up or redrawn, and the hotspots also need to be checked.


# MouseCursorPack

![mouse_cursor_pack_banner](https://user-images.githubusercontent.com/23288188/179344010-81e65575-7c4f-4348-8462-37c217b2e422.png)


This is a set of mouse cursors that can be used with the [LÖVE Framework](https://love2d.org).


## Aren't the system cursors good enough?

On the whole, yes, they are fine. CursorPack was started for a few reasons:

1) On Linux, if you mix custom cursors with system cursors, they are not guaranteed to be of the same scale.

2) On some platforms, certain system cursors obtained with [`love.mouse.getSystemCursor()`](https://love2d.org/wiki/love.mouse.getSystemCursor) are duplicates. For example, on my system, the diagonal size-arrow cursors are duplicates of sizeall.

3) A desire to have paint and adventure game-style cursors handy for future projects.

All cursors in this pack are [CC0](https://creativecommons.org/publicdomain/zero/1.0/), and can be used as the basis for new custom cursors.


## Structure

This repo contains both the finalized cursors, and work versions that can be exported with a script. The final art can be found in the `cursors` directory, arranged into subfolders by size class, and then broad category:

* **system**: Covers the system cursors you can get through `love.mouse.getSystemCursor()`. Can be used to implement fallback texture-based cursors.

* **system_alternatives**: Some variations on the cursors in `system`.

* **auxiliary_pointers**: Alternative pointer cursors which don't have `waitarrow` variations.

* **paint**: Intended for paint applications.

* **senses**: Body parts associated with external human senses.

* **object_metaphors**: Objects that can represent UI actions.


`wip-raster` and `wip-vector` contain *work* versions of the cursors. The raster images don't have transparency set, and the vector images are in SVG format. The `build_cursor_pack.lua` script is used to finalize and export images to the `cursors` folder.

If you're not interested in (re)exporting cursors, you can safely delete `wip-raster`, `wip-svg`, and `build_cursor_pack.lua`.


## Cursor format

By default, all cursors are black-and-white PNGs, with click hotspots encoded in their filenames:

`cursor_name-hx_1_hy_0.png`

Each filename contains one hyphen which separates the cursor ID from its metadata tags.

The tags are:

* `hx_<n>`: Hotspot X position

* `hy_<n>`: Hotspot Y position


The 8x8, 12x12, 16x16, 24x24 and 32x32 cursors are hand-pixeled with cut-out alpha transparency and rough edges. Everything larger than that is exported from SVG (vector) art, with smooth edges.


## test\_cursors.lua

This is a LÖVE application / script which can load all cursors. To run, you need to have LÖVE installed and available in your CLI path. Then invoke: `love . test_cursors`

`test_cursors.lua` is still a bit fragile, and can't deal with PNGs missing within a given size class.


## Notes

The 8x8 and 12x12 cursors are mostly useless on modern desktops. They are included because they might be useful in niche / retro cases.

The SVG cursor art is drawn in Inkscape with a page size of 64x64 pixels, using rescaled versions of the 32x32 raster art as a starting point. Most lines are set to a width of 2 pixels. This is the first real SVG art I've ever done, and I am likely doing some things that aren't quite right.

Some vector objects are not quite full black-and-white. (I'm having some issues with the 'value' slider in Inkscape.)


### Baking assets

Inkscape can convert SVG to PNG from the terminal:

`inkscape -w 64 -h 64 cool_cursor.svg -o cool_cursor.png`

`build_cursor_pack.lua` will export the contents of `wip-raster` and `wip-vector` to the `cursors` folder. (Inkscape is used for the vector export, so it must be installed and accessible from the CLI as `inkscape` for that part to work.) It's currently a pretty hacky solution, and I have only successfully run it on Fedora 36 so far.

The builder script can export with hotspot details encoded within the filename, or create a secondary `.hotspot` file with the offsets.

The SVG images in `wip-vector` must have hotspot positions encoded within their filenames in the range of 0 to 63. This is multiplied and floored when `build_cursor_pack.lua` exports them to PNG.


### Problems

As of July 2022, none of these cursors have been tested in any real applications. If you encounter any issues, feel free to contact me or open an issue ticket.


### Libraries

* `build_cursor_pack.lua`:
  * [NativeFS](https://github.com/EngineerSmith/nativefs)


## License

The sample program and `build_cursor_pack.lua` are MIT. The cursors are [CC0](https://creativecommons.org/publicdomain/zero/1.0/).

MIT License

Copyright (c) 2022 RBTS

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


CHANGELOG.md
# MouseCursorPack Changelog

## v0.0.2 (2022-16-07) (Beta)

* Fixed seam in `senses/hand_grasp` and `senses/hand_grasp_invert` vector art.
* Centered hotspot in `senses/sense_nose` 16px raster art.
* Started changelog.


test_cursors.lua
-- beta


--[[
	LÖVE loader demo for CursorPack.
	Supported LÖVE versions: 11.4

	Repository: TODO

	License for included PNG and SVG cursor graphics: https://creativecommons.org/publicdomain/zero/1.0/

	License for this demo file: MIT License

	Copyright (c) 2022 RBTS

	Permission is hereby granted, free of charge, to any person obtaining a copy
	of this software and associated documentation files (the "Software"), to deal
	in the Software without restriction, including without limitation the rights
	to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
	copies of the Software, and to permit persons to whom the Software is
	furnished to do so, subject to the following conditions:

	The above copyright notice and this permission notice shall be included in all
	copies or substantial portions of the Software.

	THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
	IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
	FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
	AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
	LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
	OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
	SOFTWARE.

--]]


love.keyboard.setKeyRepeat(true)


local cursorLoad = require("cursor_load")


-- cursors -> IDs -> sizes
local cursors = {}

-- In (semi) alphabetical order
local cursors_seq = {}

local show_mouse_xy = false


--[[
	NOTE: This script assumes that the cursors have been exported with hotspot tags embedded into the filenames.
--]]

local function _recursiveEnumerate(folder, _file_list, _folder_list)
	_file_list = _file_list or {}
	_folder_list = _folder_list or {}

	local filesTable = love.filesystem.getDirectoryItems(folder)

	for i, v in ipairs(filesTable) do
		local file = folder .. "/" .. v
		local info = love.filesystem.getInfo(file)

		if not info then
			print("ERROR: failed to get file info for: " .. file)

		else
			if info.type == "file" then
				local ext = string.match(file, ".*%.(.*)$") or ""
				table.insert(_file_list, file)

			elseif info.type == "directory" then
				table.insert(_folder_list, file)
				_recursiveEnumerate(file, _file_list, _folder_list)
			end
		end
	end

	return _file_list, _folder_list
end


local function loadFromPath(path, cursors)
	local file_paths = _recursiveEnumerate(path)

	for i, file_path in ipairs(file_paths) do
		print("Load cursor at:" .. file_path)

		if string.find(file_path, "%.png$") then

			local i_data, id, hx, hy = cursorLoad.loadTaggedFile(file_path)

			-- XXX very hacky
			local px = tonumber(string.match(file_path, "/px_(%d+)/"))

			local cursor = {}

			cursor.id = id
			cursor.px = px
			cursor.hx = hx
			cursor.hy = hy

			cursor.obj = love.mouse.newCursor(i_data, cursor.hx, cursor.hy)
			cursor.img = love.graphics.newImage(i_data)

			if not cursors[cursor.id] then
				cursors[cursor.id] = {id = cursor.id}
			end
			cursors[cursor.id][cursor.px] = cursor
		end
	end
end


-- Load all cursors
loadFromPath("cursors", cursors)


-- Make sorted list for consistent traversal.
do
	local function sort_func(a, b)
		return a.id < b.id;
	end

	-- To reduce flashing imagery, move the plain and inverted versions of cursors to opposite ends of the array.
	local c1, c2 = {}, {}
	for k, v in pairs(cursors) do
		if string.match(k, "invert") then
			table.insert(c1, v)
		else
			table.insert(c2, v)
		end
	end
	table.sort(c1, sort_func)
	table.sort(c2, sort_func)

	for i, v in ipairs(c1) do
		table.insert(cursors_seq, v)
	end
	for i, v in ipairs(c2) do
		table.insert(cursors_seq, v)
	end
end


local cursor_i = 1
local sizes = {8, 12, 16, 24, 32, 48, 64, 96, 128, 192, 256}
local size_i = 5

local hardware_mode = true


local function setCursor()
	love.mouse.setCursor(cursors_seq[cursor_i][sizes[size_i]].obj)
end

setCursor()


function love.keypressed(kc, sc)
	local reload_cursor = false

	if sc == "escape" then
		love.event.quit()

	elseif sc == "up" or sc == "w" then
		size_i = math.max(1, size_i - 1)
		setCursor()

	elseif sc == "down" or sc == "s" then
		size_i = math.min(#sizes, size_i + 1)
		setCursor()

	elseif sc == "left" or sc == "a" then
		cursor_i = cursor_i - 1
		if cursor_i < 1 then
			cursor_i = #cursors_seq
		end
		setCursor()

	elseif sc == "right" or sc == "d" then
		cursor_i = cursor_i + 1
		if cursor_i > #cursors_seq then
			cursor_i = 1
		end		
		setCursor()

	elseif sc == "space" then
		show_mouse_xy = not show_mouse_xy

	elseif sc == "tab" then
		hardware_mode = not hardware_mode
		love.mouse.setVisible(hardware_mode)
	end
end


--function love.update(dt)


function love.draw()
	local w, h = love.graphics.getDimensions()
	local mx, my = love.mouse.getPosition()

	local steps = 8
	local step_w = w / steps

	for i = 0, steps - 1 do
		local color = i / steps
		love.graphics.setColor(color, color, color, 1)
		love.graphics.rectangle("fill", i * step_w, 0, step_w, h)
	end

	if show_mouse_xy then
		love.graphics.setColor(1, 0.25, 0.25, 1)
		love.graphics.setLineStyle("rough")
		love.graphics.setLineWidth(1)
		love.graphics.line(0, my, w, my)
		love.graphics.line(mx, 0, mx, h)
	end

	love.graphics.setColor(0, 0, 0, 0.8)
	love.graphics.rectangle("fill", 0, h - 64, w, 64)
	love.graphics.setColor(1, 1, 1, 1)
	love.graphics.print(
		"up/down w/s: change size: " .. sizes[size_i]
		.. "\nleft/right a/d: change cursor: " .. cursors_seq[cursor_i].id
		.. "\nspace: show mouse XY: " .. tostring(show_mouse_xy)
		.. "\ntab: hardware mode: " .. tostring(hardware_mode),
		8, h - 64 + 8)

	love.graphics.print(
		"Escape: quit",
		400, h - 64 + 8)

	if not hardware_mode then
		love.graphics.setColor(1,1,1,1)
		local cursor = cursors_seq[cursor_i][sizes[size_i]]
		love.graphics.draw(cursor.img, mx - cursor.hx, my - cursor.hy)
	end
end


local function testLoadAllCursors()
	for i, cursor in ipairs(cursors_seq) do
		for j, size in ipairs(sizes) do
			print("Test load:", cursor.id, size)
			cursor_i = i
			size_i = j

			setCursor()
		end
	end
end

-- Uncomment to force-load all cursors at application start.
--testLoadAllCursors()



cursor_load.lua
--[[
	CursorLoad -- part of the CursorMgr package
	License: Source code is MIT, demo cursor art is CC0
--]]


local cursorLoad = {}


-- * Internal *


local function getIntProperty(str, id, strict)

	local retval

	local i, j = string.find(str, id)
	if i then
		retval = tonumber(string.match(str, "%d+", j + 1))
	end

	if not retval and strict then
		error("missing cursor tag: " .. id)
	end

	return retval
end


-- * / Internal *


--- Utility function to parse the cursor ID and hotspot tags within filenames.
-- @param file_path Filename, optionally including the path (which will be trimmed) (forward slashes as separators only).
-- @return The ID, X hotspot and Y hotspot parsed from the file path.
function cursorLoad.getTagsFromFilePath(file_path)

	-- Remove leading path, if applicable.
	local id = string.match(file_path, "/*([^/]*)$")

	-- Separate ID and tagged regions
	local tag_str = string.match(id, "%-(.*)%..*$")

	id = string.match(id, "([^%-]*)%-")

	if not tag_str then
		error("couldn't extract tag substring from file path.")
	end

	local hx = getIntProperty(tag_str, "hx_", true)
	local hy = getIntProperty(tag_str, "hy_", true)

	return id, hx, hy
end


--- Load an image with tagged cursor metadata.
-- @param file_path Path and name of the image file.
-- @return ImageData created from the file path, and the ID, hotspot X and hotspot Y positions parsed from the file.
function cursorLoad.loadTaggedFile(file_path)

	local id, hx, hy = cursorLoad.getTagsFromFilePath(file_path)
	local i_data = love.image.newImageData(file_path)

	return i_data, id, hx, hy
end


--- Load an image with a paired .hotspot file containing cursor metadata.
-- @param file_path Path and name of the image file.
-- @param hotspot_path If the .hotspot file is not located in the same path as 'file_path', specify it here. Otherwise, leave this blank.
-- @return ImageData created from the file path, and the ID, hotspot X and hotspot Y positions parsed from the file.
function cursorLoad.loadFilePair(file_path, hotspot_path)

	local i_data = love.image.newImageData(file_path)

	hotspot_path = hotspot_path or (string.match(file_path, "(.*)%..*$") .. ".hotspot")

	-- Remove leading path, if applicable.
	local id = string.match(file_path, "/*([^/]*)$")
	id = string.match(id, "([^%.]*)%.")

	local hs_contents, size_or_err = love.filesystem.read("string", hotspot_path)

	if not hs_contents then
		error("failed to load .hotspot file: " .. tostring(size_or_err))
	end

	local hx = getIntProperty(hs_contents, "hx ", true)
	local hy = getIntProperty(hs_contents, "hy ", true)

	return i_data, id, hx, hy
end


return cursorLoad


build_cursor_pack.lua
--[[
	WARNING: This script is incomplete and hasn't been thoroughly tested. See 'TODO' notes below.

	Exports SVG cursors, applies cut-out alpha transparency to raster images, and sorts all
	images by size class.

	Inkscape is required for exporting SVG images. Despite the love.filesystem codepath stuff,
	NativeFS is currently required.

	To run:

	LÖVE 11.4:
	love . build_cursor_pack

	LÖVE 12.0: (untested)
	love build_cursor_pack.lua
--]]

--[[
	TODO: Could this be rewritten as a Python script? Even though it was designed to make cursors for LÖVE,
	there technically isn't anything here that actually requires it.

	TODO: Do something about the Inkscape command hanging the main thread.

	TODO: This code, particularly the directory-handling logic, was written hastily and is pretty fragile.

	TODO: Test love filesystem code path

	TODO: Test on Windows
		-> Raster export works, but I haven't tested with Inkscape installed yet.
--]]


local time_start = love.timer.getTime()

-- Helps parse filenames
local cursorLoad = require("cursor_load")

-- Libraries
local nativefs = require("lib.nativefs")

--[[
	Size classes are hard-coded for now.
--]]

local size_raster = {8, 12, 16, 24, 32,}
local size_vector = {48, 64, 96, 128, 192, 256,}
local size_all = {}
for i, size in ipairs(size_raster) do
	table.insert(size_all, size)
end
for i, size in ipairs(size_vector) do
	table.insert(size_all, size)
end

-- * Filesystem wrappers * --

local fs_backend = "nfs"
local build_hs_mode = "encode_in_file" -- "encode_in_file", "write_hotspot_file"
--build_hs_mode = "write_hotspot_file"

local function fs_getBaseDir()
	if fs_backend == "lfs" then
		return ""
	else
		return nativefs.getWorkingDirectory() .. "/"
	end
end


local function fs_getInfo(file)
	if fs_backend == "lfs" then
		return love.filesystem.getInfo(file)
	elseif fs_backend == "nfs" then
		return nativefs.getInfo(file)
	else
		error("invalid filesystem backend: " .. fs_backend)
	end
end


local function fs_getDirectoryItems(folder)
	if fs_backend == "lfs" then
		return love.filesystem.getDirectoryItems(folder)
	elseif fs_backend == "nfs" then
		return nativefs.getDirectoryItems(folder)
	else
		error("invalid filesystem backend: " .. fs_backend)
	end
end


local function fs_newFileData(path)
	local f_data, err
	if fs_backend == "lfs" then
		f_data, err = love.filesystem.newFileData(path)
	elseif fs_backend == "nfs" then
		f_data, err = nativefs.newFileData(path)
	else
		error("invalid filesystem backend: " .. fs_backend)
	end

	if not f_data then
		error(err)
	else
		return f_data
	end
end


local function fs_remove(path)
	local ok, err

	if fs_backend == "lfs" then
		ok, err = love.filesystem.remove(path)
	elseif fs_backend == "nfs" then
		ok, err = nativefs.remove(path)
	else
		error("invalid filesystem backend: " .. fs_backend)
	end

	if not ok then
		error("failed to delete '" .. tostring(path) .. "': " .. tostring(err))
	end
end


local function fs_createDir(path)
	local ok, err

	if fs_backend == "lfs" then
		ok, err = love.filesystem.createDirectory(path)
	elseif fs_backend == "nfs" then
		ok, err = nativefs.createDirectory(path)
	end

	if not ok then
		error(err or "failed to create directory: " .. path) -- lf.createDirectory doesn't return an error string on failure
	end
end


local function _recursiveEnumerate(folder, _file_list, _folder_list)
	_file_list = _file_list or {}
	_folder_list = _folder_list or {}

	local filesTable = fs_getDirectoryItems(folder)

	for i, v in ipairs(filesTable) do
		local file = folder .. "/" .. v
		local info = fs_getInfo(file)

		if not info then
			print("ERROR: failed to get file info for: " .. file)

		else
			if info.type == "file" then
				local ext = string.match(file, ".*%.(.*)$") or ""
				table.insert(_file_list, file)

			elseif info.type == "directory" then
				table.insert(_folder_list, file)
				_recursiveEnumerate(file, _file_list, _folder_list)
			end
		end
	end

	return _file_list, _folder_list
end


local function _recursiveDelete(folder)
	local files, folders = _recursiveEnumerate(folder)
	for i, file in ipairs(files) do
		fs_remove(file)
	end
	
	for i = #folders, 1, -1 do
		fs_remove(folders[i])
	end
end


-- * / Filesystem wrappers * --


local base_dir = fs_getBaseDir()
local dir_cursors = base_dir .. "cursors"

-- Delete contents of 'cursors'. If you are making changes to the script, back up
-- the last known good output before running.
_recursiveDelete(base_dir .. "cursors")


-- Get directory details
local raster_files, raster_folders = _recursiveEnumerate(base_dir .. "wip-raster")
local vector_files, vector_folders = _recursiveEnumerate(base_dir .. "wip-vector")

-- Add folders to 'cursors' for every folder found in the above calls.
local function makeFolders(folders, base_len)
	for i, dir in ipairs(folders) do
		local trimmed_dir = string.sub(dir, base_len)

		for j, size in ipairs(size_all) do
			local dest_dir = dir_cursors .. "/px_" .. tostring(size) .. "/" .. trimmed_dir
			local info = fs_getInfo(dest_dir)
			if not info or (info and info.type ~= "directory") then
				fs_createDir(dest_dir)
			end
		end
	end
end
makeFolders(raster_folders, #base_dir + #"wip-raster/" + 1)
makeFolders(vector_folders, #base_dir + #"wip-vector/" + 1)


-- Load raster images, apply cut-out alpha transparency, and save to destination
for i, i_file in ipairs(raster_files) do
	local f_data = fs_newFileData(i_file)
	local i_data = love.image.newImageData(f_data)

	for y = 0, i_data:getHeight() - 1 do
		for x = 0, i_data:getWidth() - 1 do
			local pr, pg, pb = i_data:getPixel(x, y)
			if pr == 1 and pg == 0 and pb == 1 then
				i_data:setPixel(x, y, 0, 0, 0, 0)
			end
		end
	end

	local id, hx, hy = cursorLoad.getTagsFromFilePath(i_file)

	-- XXX pretty hacky
	local px = string.match(i_file, "%-px_(%d+)_")

	local path_out = string.sub(i_file, #base_dir + #"wip-raster/" + 1)
	path_out = string.match(path_out, "(.*)/[^/]*$")
	
	local file_out
	local file_hs_out
	if build_hs_mode == "encode_in_file" then
		file_out = id .. "-hx_" .. hx .. "_hy_" .. hy .. ".png"
	else
		file_out = id .. ".png"
		file_hs_out = id .. ".hotspot"
	end

	local final_out = dir_cursors .. "/px_" .. tostring(px) .. "/" .. path_out .. "/" .. file_out

	print("RASTER: write " .. final_out)

	local out_data = i_data:encode("png")
	nativefs.write(final_out, out_data)

	if build_hs_mode == "write_hotspot_file" then
		local hs_str = "hx " .. hx .. "\nhy " .. hy
		nativefs.write(dir_cursors .. "/px_" .. tostring(px) .. "/" .. path_out .. "/" .. file_hs_out, hs_str)
	end
end


-- Export SVGs with Inkscape.
for i, i_file in ipairs(vector_files) do

	-- Separate ID and tags from filename.
	-- We will scale and floor the hotspot for every exported size.
	local id, hx, hy = cursorLoad.getTagsFromFilePath(i_file)

	-- XXX pretty hacky. This is already expected to be 64.
	local px = string.match(i_file, "%-px_(%d+)_")

	hx = hx / px
	hy = hy / px

	for s, size in ipairs(size_vector) do
		local path_out = string.sub(i_file, #base_dir + #"wip-vector/" + 1)
		path_out = string.match(path_out, "(.*)/[^/]*$")

		local name_out
		local file_hs_out
		if build_hs_mode == "encode_in_file" then
			name_out = id .. "-hx_" .. math.floor(hx * size) .. "_hy_" .. math.floor(hy * size) .. ".png"
		else
			name_out = id .. ".png"
			file_hs_out = id .. ".hotspot"
		end

		local final_out = dir_cursors .. "/px_" .. size .. "/" .. path_out .. "/" .. name_out
		print("VECTOR: write " .. final_out)
		local command = "inkscape -w " .. size .. " -h " .. size .. " " .. i_file .. " -o " .. final_out

		local result = os.execute(command)

		if build_hs_mode == "write_hotspot_file" then
			local hs_str = "hx " .. math.floor(hx * size) .. "\nhy " .. math.floor(hy * size)
			nativefs.write(dir_cursors .. "/px_" .. size .. "/" .. path_out .. "/" .. file_hs_out, hs_str)
		end

		if not result then
			error("inkscape command failed: " .. command)
		end
		love.timer.sleep(0.01)
	end
end

print("\n\nAll done! Time in seconds: " .. love.timer.getTime() - time_start .. "\n\n")

love.event.quit()

